src/Zynga/Framework/Service/V2/Handler/Http.hh

```diff
@@ -86,6 +86,7 @@ class Http extends BaseHandler {
         // var_dump($data);
         // Bring the input into the dataset.
         $service->request()->import()->fromMap($data);
+        $service->server()->import()->fromMap(SuperGlobals::getServerAsMap());
```

Comment:
What's the motivation to make server a mutable storable object as opposed to a ImmutableMap ?

Reply:
The idea is that rather than give all services full access, if a service wants to use a server superglobal, the developer would have to explicitly extend and define those variables. I'm not sure that an immutable map would work in the same fashion.

Reply:
So, this is interesting. This particular function will return the raw global as a map (which makes sense).

However, if a developer is accessing the raw global as a map then they are bypassing some of the security protocols that we put in place. If you see something like StringBox (https://github.com/zynga/zynga-hacklang-framework/blob/master/src/Zynga/Framework/Type/V1/StringBox.hh) it has internally built security enforcement. 

I would be very careful about introducing this as it exposes us to skipping over security protocols. What do you have in mind for making sure that having a global like this can be accessed? 

Here is a good reference to see what I mean.
https://security.stackexchange.com/questions/32299/is-server-a-safe-source-of-data-in-php
```
Server controlled

These variables are set by the server environment and depend entirely on the server configuration.

    'GATEWAY_INTERFACE'
    'SERVER_ADDR'
    'SERVER_SOFTWARE'
    'DOCUMENT_ROOT'
    'SERVER_ADMIN'
    'SERVER_SIGNATURE'

Partly server controlled

These variables depend on the specific request the client sent, but can only take a limited number of valid values, since all invalid values should be rejected by the web server and not cause the invocation of the script to begin with. Hence they can be considered reliable.

    'HTTPS'
    'REQUEST_TIME'
    'REMOTE_ADDR' *
    'REMOTE_HOST' *
    'REMOTE_PORT' *
    'SERVER_PROTOCOL'
    'HTTP_HOST' †
    'SERVER_NAME' †
    'SCRIPT_FILENAME'
    'SERVER_PORT'
    'SCRIPT_NAME'

* The REMOTE_ values are guaranteed to be the valid address of the client, as verified by a TCP/IP handshake. This is the address where any response will be sent to. REMOTE_HOST relies on reverse DNS lookups though and may hence be spoofed by DNS attacks against your server (in which case you have bigger problems anyway). This value may be a proxy, which is a simple reality of the TCP/IP protocol and nothing you can do anything about.

† If your web server responds to any request regardless of HOST header, this should be considered unsafe as well. See How safe is $_SERVER[“HTTP_HOST”]?.
Also see http://shiflett.org/blog/2006/mar/server-name-versus-http-host.
Entirely arbitrary user controlled values

These values are not checked at all and do not depend on any server configuration, they are entirely arbitrary information sent by the client.

    'argv', 'argc' (only applicable to CLI invocation, not usually a concern for web servers)
    'REQUEST_METHOD' ‡
    'QUERY_STRING'
    'HTTP_ACCEPT'
    'HTTP_ACCEPT_CHARSET'
    'HTTP_ACCEPT_ENCODING'
    'HTTP_ACCEPT_LANGUAGE'
    'HTTP_CONNECTION'
    'HTTP_REFERER'
    'HTTP_USER_AGENT'
    'AUTH_TYPE' §
    'PHP_AUTH_DIGEST' §
    'PHP_AUTH_USER' §
    'PHP_AUTH_PW' §
    'PATH_INFO'
    'ORIG_PATH_INFO'
    'REQUEST_URI' (may contain tainted data)
    'PHP_SELF' (may contain tainted data i.e. /index.php/evilstring)
    'PATH_TRANSLATED'
    any other 'HTTP_' value

‡ May be considered reliable as long as the web server allows only certain request methods.

§ May be considered reliable if authentication is handled entirely by the web server.

The superglobal $_SERVER also includes several environment variables. Whether these are "safe" or not depend on how (and where) they are defined. They can range from completely server controlled to completely user controlled.
```

Reply:
The Importer will populate the values in the extended Server class, setting the value of any boxes matching the keys in the $_SERVER map, anything else returned from `getServerAsMap()` is discarded.

see: https://github.com/zynga/zynga-hacklang-framework/pull/47/files#diff-291cc7f40104ec66ced746844856b0f3
and https://github.com/zynga/zynga-hacklang-framework/blob/master/src/Zynga/Framework/StorableObject/V1/Object/Importer.hh#L103:L104

Reply:
I understand that the Service itself will have this imported so that the globals will be safe here. That's not the problem I was trying to point out.

The issue exists in the SuperGlobals class. That publicly exposes the functionality. 

Service Requests will have a safe version of accessing globals but other code will not if we keep this pattern. Do you see what I mean?

Reply:
@stevenramirezgames 
I don't fully follow. Lets chat in person tomorrow to reduce the back and forth here. 

As i see it right now, the SuperGlobals is simply a wrapper to read PHP globals like $_SERVER, $_POST, $_GET, etc. 

Its up to to us on how and where we want to use those. I understand that not all the properties of the $_SERVER can be fully trusted. The same applies for $_POST or $_GET. The way these properties gets used is up to the developer designing the service and the link you shared is very helpful in understanding which properties are can't be modified from the client side. 

I think we should comment that link in the base service class to that the developer knows which properties can't be spoofed. 

In terms of the security protocols, we will have to use the boxes we designed or make new boxes if an existing one can't be used. 

For example, i would use StringBox for parsing SERVER_SOFTWARE. But it might make sense to make a completely new box with its own security protocols for REQUEST_URI. 
```
Service extends BaseService {
 public StringBox SERVER_SOFTWARE;  // existing protocols are good enough here
 public URIBox REQUEST_URI; // a new box with its own security protocols
}
```

Reply:
Sure yeah, let's chat in person.

What I was trying to point out is the separation of validation of untrusted data from a data provider. You have Services fetching data in a sanitized format where other outside code would not have the same contract. Thus, SuperGlobals is now a potential security hole in an application because it does not prevent the user from reading untrusted data in a safe way.

Reply:
We talked offline. 

To keep things clean and in sync, we are using the same pattern used by request interface. The changes done here is to support reading ServerGlobals from SuperGlobals for the services framework. 

Other outside code is not yet supported and that is outside the scope of the change set in this PR.